File: Application/src/main/java/com/shareyourproxy/app/fragment/MainDrawerFragment.java
package com.shareyourproxy.app.fragment;

import android.os.Bundle;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.Toast;

import com.shareyourproxy.R;
import com.shareyourproxy.api.rx.JustObserver;
import com.shareyourproxy.api.rx.command.eventcallback.LoggedInUserUpdatedEventCallback;
import com.shareyourproxy.api.rx.event.SelectDrawerItemEvent;
import com.shareyourproxy.app.AggregateFeedActivity;
import com.shareyourproxy.app.adapter.BaseRecyclerView;
import com.shareyourproxy.app.adapter.BaseViewHolder.ItemLongClickListener;
import com.shareyourproxy.app.adapter.DrawerAdapter;
import com.shareyourproxy.app.adapter.DrawerAdapter.DrawerItem;

import butterknife.Bind;
import butterknife.ButterKnife;
import rx.subscriptions.CompositeSubscription;


/**
 * Drawer Fragment to handle displaying a user profile with options.
 */
public class MainDrawerFragment extends BaseFragment implements ItemLongClickListener {

    @Bind(R.id.fragment_drawer_recyclerview)
    BaseRecyclerView drawerRecyclerView;
    private DrawerAdapter _adapter;
    private CompositeSubscription _subscriptions;

    /**
     * Constructor.
     */
    public MainDrawerFragment() {
    }

    /**
     * Create a new instance of this fragment for parent {@link AggregateFeedActivity}.
     *
     * @return drawer fragment
     */
    public static MainDrawerFragment newInstance() {
        return new MainDrawerFragment();
    }

    @Override
    public View onCreateView(
        LayoutInflater inflater, ViewGroup container, Bundle
        savedInstanceState) {
        View view = inflater.inflate(R.layout.fragment_drawer, container, false);
        ButterKnife.bind(this, view);
        initializeRecyclerView();
        return view;
    }

    @Override
    public void onResume() {
        super.onResume();
        _subscriptions = new CompositeSubscription();
        _subscriptions.add(getRxBus().toObservable()
            .subscribe(getBusObserver()));
    }

    public JustObserver<Object> getBusObserver() {
        return new JustObserver<Object>() {
            @Override
            public void next(Object event) {
                if (event instanceof LoggedInUserUpdatedEventCallback) {
                    _adapter.updateUser(((LoggedInUserUpdatedEventCallback) event).user);
                }
            }
        };
    }

    @Override
    public void onPause() {
        super.onPause();
        _subscriptions.unsubscribe();
        _subscriptions = null;
    }

    /**
     * Initialize a recyclerView with User data and menu options.
     */
    private void initializeRecyclerView() {
        _adapter = DrawerAdapter.newInstance(getLoggedInUser(), this);

        drawerRecyclerView.setLayoutManager(new LinearLayoutManager(getActivity()));
        drawerRecyclerView.setHasFixedSize(true);
        drawerRecyclerView.setItemAnimator(new DefaultItemAnimator());
        drawerRecyclerView.setAdapter(_adapter);
    }

    @Override
    public void onItemClick(View view, int position) {
        DrawerItem drawerItem = _adapter.getSettingValue(position);
        getRxBus().post(new SelectDrawerItemEvent(
            drawerItem, view, position, getString(drawerItem.getLabelRes())));
    }

    @Override
    public void onItemLongClick(View view, int position) {
        DrawerItem item = _adapter.getSettingValue(position);
        if (!item.equals(DrawerItem.HEADER)) {
            Toast.makeText(getActivity(), getString(item.getLabelRes()), Toast.LENGTH_SHORT)
                .show();
        }
    }
}

File: Application/src/main/kotlin/com/shareyourproxy/app/fragment/MainDrawerFragment.kt
package com.shareyourproxy.app.fragment

import android.os.Bundle
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.Toast
import com.shareyourproxy.R
import com.shareyourproxy.R.id.fragment_drawer_recyclerview
import com.shareyourproxy.api.rx.JustObserver
import com.shareyourproxy.api.rx.RxBusRelay.post
import com.shareyourproxy.api.rx.RxBusRelay.rxBusObservable
import com.shareyourproxy.api.rx.command.eventcallback.LoggedInUserUpdatedEventCallback
import com.shareyourproxy.api.rx.event.SelectDrawerItemEvent
import com.shareyourproxy.app.adapter.BaseRecyclerView
import com.shareyourproxy.app.adapter.BaseViewHolder.ItemLongClickListener
import com.shareyourproxy.app.adapter.DrawerAdapter
import com.shareyourproxy.app.adapter.DrawerAdapter.DrawerItem
import com.shareyourproxy.util.ButterKnife.bindView
import rx.subscriptions.CompositeSubscription


/**
 * Drawer Fragment to handle displaying a user profile with options.
 */
internal final class MainDrawerFragment() : BaseFragment(), ItemLongClickListener {

    private val drawerRecyclerView: BaseRecyclerView by bindView(fragment_drawer_recyclerview)
    private val adapter: DrawerAdapter = DrawerAdapter(loggedInUser, this)
    private val subscriptions: CompositeSubscription = CompositeSubscription()
    private val busObserver: JustObserver<Any> = object : JustObserver<Any>(MainDrawerFragment::class.java) {
        @Suppress("PARAMETER_NAME_CHANGED_ON_OVERRIDE")
        override fun next(event: Any) {
            if (event is LoggedInUserUpdatedEventCallback) {
                adapter.updateUser(event.user)
            }
        }
    }

    override fun onCreateView(inflater: LayoutInflater, container: ViewGroup?, savedInstanceState: Bundle?): View? {
        val view = inflater.inflate(R.layout.fragment_drawer, container, false)
        initializeRecyclerView()
        return view
    }

    override fun onResume() {
        super.onResume()
        subscriptions.add(rxBusObservable().subscribe(busObserver))
    }

    override fun onPause() {
        super.onPause()
        subscriptions.unsubscribe()
    }

    override fun onItemClick(view: View, position: Int) {
        val drawerItem = adapter.getSettingValue(position)
        post(SelectDrawerItemEvent(drawerItem, view, position, getString(drawerItem.labelRes)))
    }

    override fun onItemLongClick(view: View, position: Int) {
        val item = adapter.getSettingValue(position)
        if (item != DrawerItem.HEADER) {
            Toast.makeText(activity, getString(item.labelRes), Toast.LENGTH_SHORT).show()
        }
    }

    /**
     * Initialize a recyclerView with User data and menu options.
     */
    private fun initializeRecyclerView() {
        drawerRecyclerView.layoutManager = LinearLayoutManager(activity)
        drawerRecyclerView.setHasFixedSize(true)
        drawerRecyclerView.itemAnimator = DefaultItemAnimator()
        drawerRecyclerView.adapter = adapter
    }
}
